src/assets/TokenDetectionController.ts

```diff
@@ -53,87 +57,157 @@ export class TokenDetectionController extends BaseController<
    * @param options.getTokenListState - Gets the current state of the TokenList controller.
    * @param options.getTokensState - Gets the current state of the Tokens controller.
+   * @param options.getNetworkState - Gets the state of the network controller.
+   * @param options.getPreferencesState - Gets the state of the preferences controller.
    * @param config - Initial options used to configure this controller.
    * @param state - Initial state to set on this controller.
    */
   constructor(
     {
-      onTokensStateChange,
       onPreferencesStateChange,
       onNetworkStateChange,
+      onTokenListStateChange,
       getBalancesInSingleCall,
-      addTokens,
+      addDetectedTokens,
       getTokenListState,
       getTokensState,
+      getNetworkState,
+      getPreferencesState,
     }: {
-      onTokensStateChange: (
-        listener: (tokensState: TokensState) => void,
-      ) => void;
       onPreferencesStateChange: (
         listener: (preferencesState: PreferencesState) => void,
       ) => void;
       onNetworkStateChange: (
         listener: (networkState: NetworkState) => void,
       ) => void;
+      onTokenListStateChange: (
+        listener: (tokenListState: TokenListState) => void,
+      ) => void;
       getBalancesInSingleCall: AssetsContractController['getBalancesInSingleCall'];
-      addTokens: TokensController['addTokens'];
+      addDetectedTokens: TokensController['addDetectedTokens'];
       getTokenListState: () => TokenListState;
       getTokensState: () => TokensState;
+      getNetworkState: () => NetworkState;
+      getPreferencesState: () => PreferencesState;
     },
     config?: Partial<TokenDetectionConfig>,
     state?: Partial<BaseState>,
   ) {
+    const {
+      provider: { chainId: defaultChainId },
+    } = getNetworkState();
+    const { useTokenDetection: defaultUseTokenDetection } =
+      getPreferencesState();
+
     super(config, state);
     this.defaultConfig = {
       interval: DEFAULT_INTERVAL,
-      networkType: MAINNET,
       selectedAddress: '',
-      tokens: [],
+      disabled: true,
+      chainId: defaultChainId,
+      isDetectionEnabledFromPreferences: defaultUseTokenDetection,
+      isDetectionEnabledForNetwork:
+        isTokenDetectionSupportedForNetwork(defaultChainId),
+      ...config,
     };
+
     this.initialize();
     this.getTokensState = getTokensState;
     this.getTokenListState = getTokenListState;
-    this.addTokens = addTokens;
-    onTokensStateChange(({ tokens }) => {
-      this.configure({ tokens });
+    this.addDetectedTokens = addDetectedTokens;
+    this.getBalancesInSingleCall = getBalancesInSingleCall;
+
+    onTokenListStateChange(({ tokenList }) => {
+      const {
+        disabled,
+        isDetectionEnabledForNetwork,
+        isDetectionEnabledFromPreferences,
+      } = this.config;
+      const hasTokens = Object.keys(tokenList).length;
+
+      if (
+        !disabled &&
+        isDetectionEnabledForNetwork &&
```

Comment:
It looks like these branches are not covered. There is no test for where `onTokenListStateChange` is triggered but one of these three things is `false`. Just the `disabled` case is covered.

To simplify this, maybe we could reduce this condition just to `hasTokens`, and ensure that is covered by a test? The other three are checked already at the start of `detectTokens`